File: prob2.py
# Time Complexity : O(M*N) as we go through entire matrix
# Space Complexity : O(M*N) similar to rotting oranges, 
# in worst case all element points in queue
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# Your code here along with comments explaining your approach

# using bfs put all zeros in queue and update the cells with -1 to
#  the current cell value + 1


from collections import deque
class Solution:

    def updateMatrix(self, mat):

        if len(mat) == 0 or mat is None:
            return mat
        
        q = deque()
        m, n = len(mat), len(mat[0])

        for i in range(m):
            for j in range(n):

                if mat[i][j] == 0:
                    q.append([i, j])
                else:
                    mat[i][j] = -1
        
        dirs = [[0, 1], [0, -1], [-1, 0], [1, 0]]

        while len(q) != 0:
            curr = q.popleft()
            for d in dirs:
                r = curr[0] + d[0]
                c = curr[1] + d[1]
                if  r >= 0 and c >= 0 and r < m and c < n and mat[r][c] == -1:
                    mat[r][c] = mat[curr[0]][curr[1]] + 1
                    q.append([r, c])
        return mat

File: prob2.java
// Time Complexity : O(m*n)
// Space Complexity : O(m*n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : yes, in bfs solution, forgot to add children to queue

// in bfs solution, if mat[i][j] == 0, add to queue, else make it -1
// while iterating over queue, check all 4 neighbours of current cell and distance of neighbour cell
// is distance at current cell + 1
// then add all neighbours to queue

import java.util.Queue;

public class prob2 {
    public int[][] updateMatrix(int[][] mat) {
        if(mat == null || mat.length == 0) return mat;
        Queue<int[]> q = new LinkedList<>();
        int m = mat.length; int n = mat[0].length;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(mat[i][j] == 0) q.add(new int[]{i, j});
                else mat[i][j] *= -1;
            }
        }

        int[][] dirs = {{-1,0}, {0, -1}, {1, 0}, {0, 1}};
        while(!q.isEmpty()){
            int[] curr = q.poll();
            for(int[] dir: dirs){
                int nr = curr[0] + dir[0];
                int nc = curr[1] + dir[1];
                if(nr >= 0 && nr < m && nc >= 0 && nc < n && mat[nr][nc] == -1){
                    mat[nr][nc] = mat[curr[0]][curr[1]] + 1;
                    q.add(new int[]{nr, nc});
                }
            }
        }
        return mat;
    }
    
}
